packages/contracts/src/contracts/current/protocol/Exchange/MixinMatchOrders.sol

```diff
@@ -0,0 +1,285 @@
+    SafeMath,
+    LibBytes,
+    LibStatus,
+    LibOrder,
+    LibPartialAmount,
+    MExchangeCore,
+    MMatchOrders,
+    MSettlement,
+    MTransactions
+    {
+
+    /// @dev Validates context for matchOrders. Succeeds or throws.
+    /// @param leftOrder First order to match.
+    /// @param rightOrder Second order to match.
+    function validateMatchOrdersContextOrRevert(
+        Order memory leftOrder,
+        Order memory rightOrder)
+        internal
+    {
+        // The leftOrder maker asset must be the same as the rightOrder taker asset.
+        require(areBytesEqual(leftOrder.makerAssetData, rightOrder.takerAssetData));
+
+        // The leftOrder taker asset must be the same as the rightOrder maker asset.
+        require(areBytesEqual(leftOrder.takerAssetData, rightOrder.makerAssetData));
+
+        // Make sure there is a positive spread.
+        // There is a positive spread iff the cost per unit bought (OrderA.MakerAmount/OrderA.TakerAmount) for each order is greater
+        // than the profit per unit sold of the matched order (OrderB.TakerAmount/OrderB.MakerAmount).
+        // This is satisfied by the equations below:
+        // <leftOrder.makerAssetAmount> / <leftOrder.takerAssetAmount> >= <rightOrder.takerAssetAmount> / <rightOrder.makerAssetAmount>
+        // AND
+        // <rightOrder.makerAssetAmount> / <rightOrder.takerAssetAmount> >= <leftOrder.takerAssetAmount> / <leftOrder.makerAssetAmount>
+        // These equations can be combined to get the following:
+        require(safeMul(leftOrder.makerAssetAmount, rightOrder.makerAssetAmount) >= safeMul(leftOrder.takerAssetAmount, rightOrder.takerAssetAmount));
+    }
+
+    /// @dev Calculates fill amounts for the matched orders.
+    ///      Each order is filled at their respective price point. However, the calculations are
+    ///      carried out as though the orders are both being filled at the right order's price point.
+    ///      The profit made by the leftOrder order goes to the taker (who matched the two orders).
+    /// @param leftOrder First order to match.
+    /// @param rightOrder Second order to match.
+    /// @param leftOrderStatus Order status of left order.
+    /// @param rightOrderStatus Order status of right order.
+    /// @param leftOrderFilledAmount Amount of left order already filled.
+    /// @param rightOrderFilledAmount Amount of right order already filled.
+    /// @return status Return status of calculating fill amounts. Returns Status.SUCCESS on success.
+    /// @return matchedFillOrderAmounts Amounts to fill left and right orders.
+    function calculateMatchedFillAmounts(
+        Order memory leftOrder,
+        Order memory rightOrder,
+        uint8 leftOrderStatus,
+        uint8 rightOrderStatus,
+        uint256 leftOrderFilledAmount,
+        uint256 rightOrderFilledAmount)
+        internal
+        returns (
+            uint8 status,
+            MatchedOrderFillAmounts memory matchedFillOrderAmounts)
```

Comment:
This should be `matchedOrderFillAmounts`, right?

Reply:
Changed this to `matchedFillResults` and `MatchedFillResults` to match the struct name (`FillResults`) in the `MixinExchangeCore`.